### ISSUE TITLE
chromium won't open fully maximized on launch

### ISSUE THREAD

**Issue by charbelnicolas** [labwc/labwc#1994]
I just updated to the latest labwc-git build and chrome is now throwing this error when launching, it wasn't happening with wlroots 0.17.

```
00:01:55.849 [ERROR] [types/xdg_shell/wlr_xdg_surface.c:169] A configure is scheduled for an uninitialized xdg_surface 0x5a15eae0d0f0
00:01:55.849 [ERROR] [types/xdg_shell/wlr_xdg_surface.c:169] A configure is scheduled for an uninitialized xdg_surface 0x5a15eae0d0f0
00:01:57.989 [ERROR] [types/xdg_shell/wlr_xdg_surface.c:169] A configure is scheduled for an uninitialized xdg_surface 0x5a15eac47720
00:01:57.989 [ERROR] [types/xdg_shell/wlr_xdg_surface.c:169] A configure is scheduled for an uninitialized xdg_surface 0x5a15eac47720
00:02:07.377 [ERROR] [types/xdg_shell/wlr_xdg_surface.c:169] A configure is scheduled for an uninitialized xdg_surface 0x5a15eae1e6a0
00:02:07.377 [ERROR] [types/xdg_shell/wlr_xdg_surface.c:169] A configure is scheduled for an uninitialized xdg_surface 0x5a15eae1e6a0
```

I noticed this because chromium won't open fully maximized on launch (a small 1/2 inch gap is left at the bottom between the display's border and chromium's border), it is only after a few milliseconds that it will restore the window to its full maximized state.

I don't know who is at fault here, if it's either chromium, labwc or wlroots... but I figured I could start reporting the issue here. THis does not happen with other programs only with chromium.

**Comment by Consolatis** [comment]
That sounds like a labwc issue, quite possible that we missed something.

**Comment by charbelnicolas** [comment]
These are the only rules I have in my config:

```
<windowRules>

	<windowRule identifier="*" serverDecoration="no" fixedPosition="yes" />

	<windowRule identifier="xdg-desktop-portal-gtk" serverDecoration="no" fixedPosition="yes">
		<action name="Maximize" />
	</windowRule>

</windowRules>
```

**Comment by charbelnicolas** [comment]
Could it have to do with server decorations? Seems the gap's height is about the same space that the server decorations would take...

**Comment by Consolatis** [comment]
Could reproduce the issue, it happens with `--start-maximized` so its unrelated to window rules or decorations.
It seems chromium is requesting maximize before the surface is being initialized (which might be totally fine). I'll look into our options.

```gdb
(gdb) break types/xdg_shell/wlr_xdg_surface.c:169
(gdb) r
[..]
Thread 1 "labwc" hit Breakpoint 1, wlr_xdg_surface_schedule_configure (surface=0xaaaaaaee7590) at ../subprojects/wlroots/types/xdg_shell/wlr_xdg_surface.c:169
169                     wlr_log(WLR_ERROR, "A configure is scheduled for an uninitialized xdg_surface %p",
(gdb) bt
#0  wlr_xdg_surface_schedule_configure (surface=0xaaaaaaee7590) at ../subprojects/wlroots/types/xdg_shell/wlr_xdg_surface.c:169
#1  0x0000aaaaaab3cc38 in wlr_xdg_toplevel_set_maximized (toplevel=0xaaaaaaee7770, maximized=true) at ../subprojects/wlroots/types/xdg_shell/wlr_xdg_toplevel.c:565
#2  0x0000aaaaaaae4da4 in xdg_toplevel_view_maximize (view=0xaaaaaaee79b0, maximized=true) at ../src/xdg.c:363
#3  0x0000aaaaaaadf714 in set_maximized (view=0xaaaaaaee79b0, maximized=VIEW_AXIS_BOTH) at ../src/view.c:1136
#4  0x0000aaaaaaadfb68 in view_maximize (view=0xaaaaaaee79b0, axis=VIEW_AXIS_BOTH, store_natural_geometry=true) at ../src/view.c:1250
#5  0x0000aaaaaaae4af4 in handle_request_maximize (listener=0xaaaaaaee7c38, data=0x0) at ../src/xdg.c:283
#6  0x0000fffff7f8a3f0 in wl_signal_emit_mutable (signal=0xaaaaaaee7858, data=0x0) at ../subprojects/wayland/src/wayland-server.c:2314
#7  0x0000aaaaaab3c4fc in xdg_toplevel_handle_set_maximized (client=0xaaaaaadf6530, resource=0xaaaaaaee7920) at ../subprojects/wlroots/types/xdg_shell/wlr_xdg_toplevel.c:356
#8  0x0000fffff69f63e0 in  () at /lib/aarch64-linux-gnu/libffi.so.8
#9  0x0000fffff69f5b24 in  () at /lib/aarch64-linux-gnu/libffi.so.8
#10 0x0000fffff7f8ff84 in wl_closure_invoke (closure=0xaaaaaaee7410, flags=2, target=0xaaaaaaee7920, opcode=9, data=0xaaaaaadf6530) at ../subprojects/wayland/src/connection.c:1228
#11 0x0000fffff7f87d00 in wl_client_connection_data (fd=30, mask=1, data=0xaaaaaadf6530) at ../subprojects/wayland/src/wayland-server.c:444
#12 0x0000fffff7f8b920 in wl_event_source_fd_dispatch (source=0xaaaaaac3be80, ep=0xffffffffe578) at ../subprojects/wayland/src/event-loop.c:113
#13 0x0000fffff7f8d17c in wl_event_loop_dispatch (loop=0xaaaaaac14360, timeout=-1) at ../subprojects/wayland/src/event-loop.c:1105
#14 0x0000fffff7f89564 in wl_display_run (display=0xaaaaaac46b40) at ../subprojects/wayland/src/wayland-server.c:1530
#15 0x0000aaaaaaacbf5c in main (argc=2, argv=0xfffffffff548) at ../src/main.c:220
```

**Comment by Consolatis** [comment]
Does #1995 improve things regarding
> I noticed this because chromium won't open fully maximized on launch (a small 1/2 inch gap is left at the bottom between the display's border and chromium's border), it is only after a few milliseconds that it will restore the window to its full maximized state.

?

**Comment by charbelnicolas** [comment]
I'm on Arch Linux, please forgive my ignorance but I don't know the exact process to apply these patches without doing a full paru -Syu update. Let me see if I can figure it out.

**Comment by Consolatis** [comment]
Should be something like
```sh
git clone https://github.com/labwc/labwc
cd labwc
git fetch origin pull/1995/head && git checkout FETCH_HEAD
meson setup build
meson compile -C build
```
Your new labwc executable should be available via `build/labwc` which you can either start manually from the TTY or nested within labwc.

**Comment by charbelnicolas** [comment]
Thank you, I followed your instructions and it did not make a difference on the maximizing part, it is still the same (I hope I did not mess it up). I ran labwc on tty2.

![IMG_20240717_194634](https://github.com/user-attachments/assets/752dceee-724a-487d-be14-98b996e4f03b)

I got some more errors though:

![ss_2024_07_17_19_48_45](https://github.com/user-attachments/assets/37090205-6163-4050-a07e-21913a5a4600)

**Comment by Consolatis** [comment]
could you do a `git describe` in your labwc repo folder?

**Comment by charbelnicolas** [comment]
Yes! I figured that out about a minute ago too hahaha... I didn't actually run the newly compiled labwc... My bad.

**Comment by charbelnicolas** [comment]
Cool! The maximized part is fixed but I still get the error about the uninitialized xdg_surface.

**Comment by Consolatis** [comment]
> Cool! The maximized part is fixed

Nice.

> but I still get the error about the uninitialized xdg_surface.

Hm.. there must be something different triggering it then. Does it happen when you only run the browser? Or only when you use `--start-maximized`?. You can test this nested as well, e.g. just starting labwc within labwc, that makes it easier to read to log output.

**Comment by charbelnicolas** [comment]
> > Cool! The maximized part is fixed
> 
> Nice.
> 
> > but I still get the error about the uninitialized xdg_surface.
> 
> Hm.. there must be something different triggering it then. Does it happen when you only run the browser? Or only when you use `--start-maximized`? You can test this nested as well, e.g. just starting labwc within labwc, that makes it easier to read to log output.

Sorry, my bad, the uninitialized xdg_surface error is gone (I was checking the system wide labwc  installation log file), now I only get the "ignoring maximize before map" warning.

**Comment by Consolatis** [comment]
> now I only get the "ignoring maximize before map" error.
> 
> ```
> 00:00:02.250 [ERROR] [../src/xdg.c:288] ignoring maximize before map
> 00:00:04.372 [ERROR] [../src/xdg.c:288] ignoring maximize before map
> ```

Yeah, those are expected, they are just for debugging. I'll remove those from the PR before we merge it.

**Comment by charbelnicolas** [comment]
Sweet, thanks for your time, I appreciate it.

**Comment by tokyo4j** [comment]
I see similar a similar message with `fuzzel` (wlr-layer-surface client):

```
00:00:02.971 [ERROR] [types/wlr_layer_shell_v1.c:296] A configure is sent to an uninitialized wlr_layer_surface_v1 0x604b6c77e060
```

I haven't investigated the cause yet.

**Comment by charbelnicolas** [comment]
I'm afraid the delayed maximization is still present :(

**Comment by charbelnicolas** [comment]
https://drive.google.com/file/d/1Cy0DAAED6AwRy_5yfGh4wasW2zC0Ih_C/view?usp=sharing

**Comment by tokyo4j** [comment]
> I'm afraid the delayed maximization is still present :(

Unfortunately I cannot reproduce it and cannot see the problem from the video. Output of `WAYLAND_DEBUG=client chromium --start-maximized` might be helpful.

**Comment by charbelnicolas** [comment]
You have to look at the bottom of the video to see that there is always a gap at the bottom at launch, it is hard to see because of the video player controls, sorry. This did not happen when using wlroots 0.17.

BTW, [https://github.com/labwc/labwc/pull/1995](https://github.com/labwc/labwc/pull/1995) did fix it, so I'm pretty sure there is a fix for it. Glitches like this are easy to spot if they were never there before.

Please check this new video, maybe it will be clearer.

[https://drive.google.com/file/d/1VxTuDIpzwc3NRW4PJdqXVEarmi74MNMc/view?usp=sharing](https://drive.google.com/file/d/1VxTuDIpzwc3NRW4PJdqXVEarmi74MNMc/view?usp=sharing)

If this doesn't happen to you, you either have a really fast machine which makes it unnoticeable or I'm doing something wrong.

**Comment by tokyo4j** [comment]
Oh I overlooked it. Thanks.

I still cannot reproduce it, so it should be dependent on personal setup.

I have no idea what's going on. Output of `WAYLAND_DEBUG=client chromium --start-maximized` will still be helpful.

**Comment by charbelnicolas** [comment]
Sure, here's the text file with the output of `WAYLAND_DEBUG=client chromium --start-maximized`

[debug.txt](https://github.com/user-attachments/files/16325851/debug.txt)

**Comment by charbelnicolas** [comment]
Using xwayland works fine but it bugs out when using chromium with these `chromium-flags.conf`:

```
--enable-vulkan
--gtk-version=4
--ozone-platform=wayland
--enable-features=UseOzonePlatform
--disable-smooth-scrolling
--enable-chrome-refresh-2023
--enable-chrome-browser-cloud-management
```

**Comment by tokyo4j** [comment]
```
[ 823374.401]  -> zxdg_toplevel_decoration_v1@38.set_mode(2)
```
This means Chromium is requesting server-side decoration, which is different from my setup. Did you set any window rule for Chromium?

**Comment by charbelnicolas** [comment]
> Could it have to do with server decorations? Seems the gap's height is about the same space that the server decorations would take...

> These are the only rules I have in my config:
> 
> ```
> <windowRules>
> 
> 	<windowRule identifier="*" serverDecoration="no" fixedPosition="yes" />
> 
> 	<windowRule identifier="xdg-desktop-portal-gtk" serverDecoration="no" fixedPosition="yes">
> 		<action name="Maximize" />
> 	</windowRule>
> 
> </windowRules>
> ```

I disabled server decorations altogether, that's what I said at the very start of the thread, that the space missing seemed about the height of server decorations hehehe

**Comment by tokyo4j** [comment]
Oh sorry I missed that comment... I tried enabling SSD from chromium side then disabling it via window rule, then I could reproduce the problem.

So at least disabling SSD by selecting "Use system bars and borders" in the Chromium's titlebar menu will fix this.

I'm not sure your issue is bug. I'll need to investigate why #1995 fixed it.

**Comment by charbelnicolas** [comment]
If I disable this:

![ss_2024_07_21_15_08_54](https://github.com/user-attachments/assets/d3b6afa6-8102-4d9e-9b33-f253a102e0ee)

The problem goes away. The funny thing is this has been disabled for the longest time and had not been an issue until wlroots 0.18.

And if I disable the system title bar and borders I only get some garbled icons instead, which is probably another issue altogether and mostly chromium related.

![ss_2024_07_21_15_11_36](https://github.com/user-attachments/assets/462cb411-cda5-434e-86dc-6bea304786bb)

**Comment by tokyo4j** [comment]
In my expectation, following things happened with the current master:

1. Chromium enables SSD and maximizes the window
2. Labwc sends the surface geometry with the area for SSD removed
3. Chromium draws its contents with that geometry
4. Labwc applies the window rule which disables SSD then sends the fullscreened surface geometry without SSD area
5. Chroimium re-draws its contents with fullscreen geometry

#1995 removes (2.) so Chromium draws its contents with the geometry that it remembers, which is fullscreen geometry.

So, I think #2007 is doing nothing wrong. To fix that issue, applying window rule for SSD at the window creation rather than at mapping will work (will be discussed in #1718).

**Comment by charbelnicolas** [comment]
Cool, I'm glad you were able to sort it out, thank you!
